### PR TITLE
Reject duplicate sorting keys

### PR DESCRIPTION
**Pull request by ulysses-you**
A SQL query may specify duplicate sorting keys:

```
SELECT c1 FROM t ORDER BY c1, c1
```

Such query fails in Gluten:

```
Caused by: java.lang.RuntimeException: Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (1 vs. 2)
Retriable: False
Expression: input_->size() >= sortCompareFlags_.size()
Function: SortBuffer
File: /opt/gluten/gluten/ep/build-velox/build/velox_ep/velox/exec/SortBuffer.cpp
Line: 36
```

Harden OrderByNode to reject duplicate sorting keys. The query planner should ensure the sorting keys are unique.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @xiaoxmeng @kewang1024 if you have time to take a look, thank you

**Comment by FelixYBW** [comment]
@rui-mo Can you review?

**Review by rui-mo — COMMENTED** [review]
Thanks for your work. Since sorting using duplicate keys yields the same results as sorting using a single key, I wonder if we can handle this at the planner level by making the keys unique. How do you think?

**Comment by ulysses-you** [comment]
It's an kind of optimization. I think it's better to do not make execution engine depends on the optimization.

**Comment by ulysses-you** [comment]
@rui-mo thank you for the thought. I think the optimization is orthogonal with this issue.

The two attributes is just a kind of duplicate key, in geneal, the duplicate key in sql query means the two canonicalized expr has the same semantics. So, if we want to optimize such case, we can add some rules to eliminate the duplicate expr for the ordering.

It also affects the Spark ordering framework since it may reduce the sort operator if it satisfies the child required ordering after optimization. At engine side, it's hard to fully optimize the query plan.

It may be better to let the engine allow duplicate keys to decouple with plan optimizer. At least, that's what Spark did...

**Review by rui-mo — COMMENTED** [review]
> It also affects the Spark ordering framework since it may reduce the sort operator if it satisfies the child required ordering after optimization.

@ulysses-you I assume we can make the keys unique when creating the Velox plan, and the Spark ordering framework could stay untouched.

**Review by rui-mo — COMMENTED** [review]
Perhaps the community could offer some advice on whether to support or reject duplicate keys in Sort. cc: @Yuhta @mbasmanova Thanks.

**Comment by Yuhta** [comment]
Looks like a performance bug to me if we are sorting on duplicated keys.  We should either eliminate it in planner (and add check in Velox), or eliminate it in Velox.  @mbasmanova What do you think?

**Comment by mbasmanova** [comment]
@Yuhta I agree with Jimmy. It would make sense for Velox to reject duplicate sorting keys.

**Comment by ulysses-you** [comment]
Thank you all for the thought.

One other small nit is that, before the behavior is a bit confused. We partial support duplicated sort keys if the input size is big engouh, e,g, `SELECT c1, c2 FROM t ORDER BY c1, c1`.

So, I changed to eliminate the duplicated sort keys when construct `OrderByNode` at logical side.

cc @rui-mo @Yuhta @mbasmanova if you have time to take another look.

File: velox/core/PlanNode.h [hunk 1]
```diff
@@ -1731,16 +1731,26 @@ class OrderByNode : public PlanNode {
...
+      : PlanNode(id), isPartial_(isPartial), sources_{source} {
     VELOX_USER_CHECK(!sortingKeys.empty(), "OrderBy must specify sorting keys");
     VELOX_USER_CHECK_EQ(
         sortingKeys.size(),
         sortingOrders.size(),
         "Number of sorting keys and sorting orders in OrderBy must be the same");
+    std::vector<FieldAccessTypedExprPtr> deduplicatedSortingKeys;
+    std::vector<SortOrder> deduplicatedSortingOrders;
+    std::unordered_set<std::string> sortingKeySet;
+    // Eliminate duplicated sorting keys.
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1743]: Instead of eliminating duplicate keys, let's just reject duplicates and document that duplicate keys are not allowed.

Comment (ulysses-you) [reply velox/core/PlanNode.h]: addressed

File: velox/core/PlanNode.h [hunk 2]
```diff
@@ -1741,6 +1741,15 @@ class OrderByNode : public PlanNode {
         sortingKeys.size(),
         sortingOrders.size(),
         "Number of sorting keys and sorting orders in OrderBy must be the same");
+    std::unordered_set<std::string> sortingKeySet;
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1744]: nit: sortingKeySet -> uniqueKeys

File: velox/core/PlanNode.h [hunk 3]
```diff
@@ -1741,6 +1741,15 @@ class OrderByNode : public PlanNode {
...
+    // Reject duplicated sorting keys.
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1745]: 1. duplicated -> duplicate
2. Move this comment before 'sortingKeySet' variable.

File: velox/core/PlanNode.h [hunk 4]
```diff
@@ -1741,6 +1741,15 @@ class OrderByNode : public PlanNode {
...
+    for (int i = 0; i < sortingKeys.size(); i++) {
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1746]: Use for-each loop.

```
for (const auto& key : sortingKeys)
```

File: velox/core/PlanNode.h [hunk 5]
```diff
@@ -1741,6 +1741,15 @@ class OrderByNode : public PlanNode {
...
+      auto filed = sortingKeys.at(i);
+      VELOX_USER_CHECK(filed, "OrderBy sorting key can not be null");
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1748]: use VELOX_USER_CHECK_NOT_NULL

Use consistent error messages: "Sorting key cannot be null"

File: velox/core/PlanNode.h [hunk 6]
```diff
@@ -1741,6 +1741,15 @@ class OrderByNode : public PlanNode {
...
+      VELOX_USER_CHECK(
+          sortingKeySet.insert(filed->name()).second,
+          "Duplicated sorting keys are not allowed: " + filed->name());
```
Comment (mbasmanova) [inline velox/core/PlanNode.h:1751]: 1. Duplicate sorting keys ...
2. Do not use string concatenation.

```
VELOX_USER_CHECK(
          sortingKeySet.insert(filed->name()).second,
          "Duplicate sorting keys are not allowed: {}", key->name());
```

File: velox/core/tests/PlanNodeTest.cpp [hunk 7]
```diff
@@ -97,3 +97,17 @@ TEST(TestPlanNode, sortOrder) {
     }
   }
 }
+
+TEST(TestPlanNode, rejectDuplicatedSortKeys) {
```
Comment (mbasmanova) [inline velox/core/tests/PlanNodeTest.cpp:101]: rejectDuplicatedSortKeys -> duplicateSortingKeys

File: velox/core/tests/PlanNodeTest.cpp [hunk 8]
```diff
@@ -97,3 +97,17 @@ TEST(TestPlanNode, sortOrder) {
...
+  auto inputSortingKeys = std::vector<FieldAccessTypedExprPtr>{
```
Comment (mbasmanova) [inline velox/core/tests/PlanNodeTest.cpp:102]: drop 'input': sortingKeys, sortingOrders

File: velox/core/tests/PlanNodeTest.cpp [hunk 9]
```diff
@@ -97,3 +97,17 @@ TEST(TestPlanNode, sortOrder) {
...
+      std::make_shared<core::FieldAccessTypedExpr>(BIGINT(), "c0"),
+      std::make_shared<core::FieldAccessTypedExpr>(BIGINT(), "c1"),
+      std::make_shared<core::FieldAccessTypedExpr>(BIGINT(), "c0"),
+  };
+  auto inputSortingOrders =
+      std::vector<SortOrder>{{true, true}, {false, false}, {true, true}};
+  EXPECT_THROW(
```
Comment (mbasmanova) [inline velox/core/tests/PlanNodeTest.cpp:109]: user VELOX_ASSERT_USER_THROW

File: velox/docs/develop/operators.rst [hunk 10]
```diff
@@ -592,7 +592,7 @@ identified sort fields as well as a sorting order.
    * - Property
      - Description
    * - sortingKeys
-     - List of one of more input columns to sort by.
+     - List of one of more input columns to sort by. Note: the duplicated sorting keys are not allowed.
```
Comment (mbasmanova) [inline velox/docs/develop/operators.rst:595]: > Note: the duplicated sorting keys are not allowed.

Sorting keys must be unique.

**Review by mbasmanova — COMMENTED** [review]
@ulysses-you Thank you for iterating. Overall looks good. Some comments below. Please, update PR description to match the code.

**Comment by ulysses-you** [comment]
thank you @mbasmanova, addressed comments.

**Review by mbasmanova — APPROVED** [review]
@ulysses-you Thanks.

**Pull request closed without merging**